### PR TITLE
Mg dm hookups

### PR DESCRIPTION
**Pull request by wence-**
Hookup DMs properly with the solvers.  This allows for reasonably transparent usage of multigrid solvers if setup.

Example in tests/multigrid/test_poisson_gmg.py

Needs update to PETSc/petsc4py which have now landed (but not yet on firedrake branch).  Reviews welcome at this point though.

### CODE REVIEW

**Comment by wence-** [comment]
All deps have now landed I think.

**Comment by wence-** [comment]
Currently missing:

- GMG on inner solves in fieldsplits, because I haven't set up the sub DMs properly
- Separate J and Jp on coarse grids (because PETSc support is not there)

**Comment by wence-** [comment]
Probably many other things besides, but those are the ones that immediately spring to mind.

File: firedrake/constant.py [hunk 1]
```diff
@@ -10,6 +10,21 @@
...
+def _globalify(value):
+    data = np.array(value, dtype=np.float64)
+    shape = data.shape
+    rank = len(shape)
+    if rank == 0:
+        dat = op2.Global(1, data)
+    elif rank == 1:
+        dat = op2.Global(shape, data)
+    elif rank == 2:
+        dat = op2.Global(shape, data)
```
Comment (kynan) [inline firedrake/constant.py]: Could combine in 1 branch

**Comment by kynan** [comment]
From a very cursory glance this looks sane to me.

**Comment by wence-** [comment]
Merged current state.  There will be more changes to come, but better to get this in first.

**Pull request merged**